### ISSUE TITLE
[Outreachy spring 2022] Applicants tasks/contributions

### ISSUE THREAD

**Issue by bluca** [systemd/systemd#22192]
# Intro

This ticket will serve as the central point tracking the contributions of perspective Outreachy (spring 2022 cohort) applicants.

For those interested, this process is described here: https://www.outreachy.org/apply/make-contributions/
The systemd project will participate for the first time this year: https://www.outreachy.org/communities/cfp/systemd/

A series of tasks will be listed here. Some of them will be "repeatable", ie: some basic tasks that every applicant can do. Some can only be done once, so they are first-come first-served. Some will be relatively quick and should be doable in a couple hours, others might take several days, and they'll be tagged as such.

Every applicant should start with the basic repeatable tasks, and complete at least two. Simply quote the task in a comment, posting proof of completion. We will "thumbs up" the comment to acknowledge completion.

After at least two of those are done and acked, then applicants should move on to the "non-repeatable" tasks. Write a comment quoting the task you selected, and if it's open-ended specify what exactly you are going to do, before you start. Those should end up with a pull request. Note that simply opening a pull request is not enough, they will be reviewed and might need follow-ups before they can be merged, so take that into account when planning for time.

Applicants should complete and wait for ack on completion at least 2 of the repeatable tasks before moving on to the short/long tasks.

@anitazha and @DaanDeMeyer kindly volunteered to help with moderating and reviewing this process with me, so look for reviews/acks from any one one of us.

# Documentation for new contributors
Contributors who are new to `systemd` are highly encouraged to start from the following documents:

- https://systemd.io/CONTRIBUTING/
- https://systemd.io/CODING_STYLE/
- https://systemd.io/HACKING/
- https://systemd.io/CODE_QUALITY/
- https://systemd.io/ARCHITECTURE/

# Repeatable tasks

- clone the project, build it, add an `Hello World from <name>!` log message to a tool of your choice from `src/`, run it and post a screenshot as proof
- clone the project, build it, pick one tool under `src/` of your choice and run it, inspect the code and in your own word write a paragraph explaining what the code (or a section of the code, some are quite large!) does - note that we don't mean just from the point of view of a user, as that's explained in the manual, but describe the actual code
- study the documentation, and write a paragraph summarizing one of the documents in your own words. Pick at least one medium/large size document, OR at least three smaller ones. For easier viewing, documents under `docs/` are rendered on https://systemd.io/ and documents under `man/` are rendered on https://www.freedesktop.org/software/systemd/man/

# Short tasks
- fix one of the issues labelled as "please-submit-as-pr" https://github.com/systemd/systemd/issues?q=is%3Aopen+is%3Aissue+label%3Aplease-submit-as-pr
- improve an existing unit test `src/test/` to increase code coverage for the basic common code - see under `src/basic/` on https://coveralls.io/github/systemd/systemd
- add a new unit test to cover the 'shared' common code - see under `src/shared/` on https://coveralls.io/github/systemd/systemd
- update outdated executables/daemons to `bus_call_method`/`bus_get_property`: https://github.com/systemd/systemd/issues/20928#issuecomment-933718629
- update outdated executables/daemons to new-style sd-bus macros: https://github.com/systemd/systemd/issues/20928#issuecomment-933733352

# Long tasks
Please note that the description might sound simple, but in reality the task might require more time than you have available. Before starting, state in a comment which issue/TODO you'd like to tackle and how much time you have available, and we'll try to advise on whether it's enough to complete it.

- fix one of the issues labelled as "RFE" https://github.com/systemd/systemd/issues?q=is%3Aopen+is%3Aissue+label%3A%22RFE+%F0%9F%8E%81%22
- fix one of the TODO items https://github.com/systemd/systemd/blob/main/TODO
- add an integration test `test/README.testsuite` for a tool/daemon that is not currently covered

**Comment by giftcup** [comment]
Hello everyone, here's my repeatable task 3. I summarized the coding style documentation.

When contributing to this codebase, use an 8ch indent for all files except the files in man/ which takes a 2ch indent and shell scripts which are indented by 4ch. When breaking a function definition into multiple lines, put all the function parameters on separate lines, using an indentation of 16ch for each of them. In all cases, do not use tabs. Input function parameters should be declared as const if they are pointers, so that they are not changed by the function and return parameters(output only) are prefixed with ret_. Also, only use the multiline comment style in code you commit. Structures should be named in PascalCase while variables and functions in snake_case. Always allocate local variables at the top of the block or where they are being used and separate variable declarations from function invocations. Take into consideration the values the variable should support before assigning types. Such as, do not use int types for variables which are always positive  but rather use unsigned. Also, do not use kernel types.  When writing infinite loops, use for(;;) instead of while(1). Lastly, commit messages should be prefixed with an appropriate component name.

**Comment by giftcup** [comment]
Hello, I've been trying to build the project for a while using meson build, but it keeps failing. Here's the error I get when I'm trying to build a particular file in src, I get an error but the entire systemd project builds. I've been unable to figure out what I am doing wrongly and I'll appreciate some help. 

![mesonerror](https://user-images.githubusercontent.com/53094535/161390758-c95827dd-43e4-4bc8-a718-6f1f46d1087a.png)

**Comment by amarjargal** [comment]
> Hello, I've been trying to build the project for a while using meson build, but it keeps failing. Here's the error I get when I'm trying to build a particular file in src, I get an error but the entire systemd project builds. I've been unable to figure out what I am doing wrongly and I'll appreciate some help.
> 
> ![mesonerror](https://user-images.githubusercontent.com/53094535/161390758-c95827dd-43e4-4bc8-a718-6f1f46d1087a.png)

Hi. You should do this in the project directory where meson files exist, in your case ~/Documents/Outreachy-systemd/systemd

**Comment by amarjargal** [comment]
Hello everyone,
For repeatable task 2, I chose the tool backlight. Running and inspecting the code, I tried to explain below what it does.

When systemd-backlight tool is run, it first checks if the argument contains "help" keyword, it prints the manual for the tool. Then it performs the following. All the untrue cases return an error message.
- checks if the number of arguments are correct (3)
- checks if the first argument contains either of the "load" or "save" keyword and if it does, creates a directory "/var/lib/systemd/backlight" if it doesn't already exist.
- checks the existence of ":" marker in the second argument and saves the name of the backlight device (which is given after ":" marker)
- checks if the seconds argument contains either of the "backlight" or "leds" keyword (type of the backlight device)
- checks if the device exists (by searching for it in "/sys/subsystem", "/sys/bus/", "/sys/class" directories)
- checks if the maximum brightness value of the backlight device is less than zero and if equal to zero. If neither of them are true, proceeds to the next step.
- If the first argument is "load", reads the brightness value of the device from "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file. If clamp option is true, it clamps (reduces) the read value. Sets the device's brightness attribute to the read value.
- If the first argument is "save", gets the brightness attribute of the device and saves the value to the "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file.

Thank you,

**Comment by bluca** [comment]
> Hello everyone, For repeatable task 2, I chose the tool backlight. Running and inspecting the code, I tried to explain below what it does.
> 
> When systemd-backlight tool is run, it first checks if the argument contains "help" keyword, it prints the manual for the tool. Then it performs the following. All the untrue cases return an error message.
> 
> * checks if the number of arguments are correct (3)
> * checks if the first argument contains either of the "load" or "save" keyword and if it does, creates a directory "/var/lib/systemd/backlight" if it doesn't already exist.
> * checks the existence of ":" marker in the second argument and saves the name of the backlight device (which is given after ":" marker)
> * checks if the seconds argument contains either of the "backlight" or "leds" keyword (type of the backlight device)
> * checks if the device exists (by searching for it in "/sys/subsystem", "/sys/bus/", "/sys/class" directories)
> * checks if the maximum brightness value of the backlight device is less than zero and if equal to zero. If neither of them are true, proceeds to the next step.
> * If the first argument is "load", reads the brightness value of the device from "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file. If clamp option is true, it clamps (reduces) the read value. Sets the device's brightness attribute to the read value.
> * If the first argument is "save", gets the brightness attribute of the device and saves the value to the "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file.
> 
> Thank you,

Thanks, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.

**Comment by giftcup** [comment]
> > Hello, I've been trying to build the project for a while using meson build, but it keeps failing. Here's the error I get when I'm trying to build a particular file in src, I get an error but the entire systemd project builds. I've been unable to figure out what I am doing wrongly and I'll appreciate some help.
> > ![mesonerror](https://user-images.githubusercontent.com/53094535/161390758-c95827dd-43e4-4bc8-a718-6f1f46d1087a.png)
> 
> Hi. You should do this in the project directory where meson files exist, in your case ~/Documents/Outreachy-systemd/systemd

Thank you for this.

**Comment by giftcup** [comment]
Hello, 
Here is my proof of completion for repeatable task 1.
![hostnamectl](https://user-images.githubusercontent.com/53094535/161615450-3574b41d-89a8-47e6-bd7f-20c01b598a39.png)

**Comment by amarjargal** [comment]
> > Hello everyone, For repeatable task 2, I chose the tool backlight. Running and inspecting the code, I tried to explain below what it does.
> > When systemd-backlight tool is run, it first checks if the argument contains "help" keyword, it prints the manual for the tool. Then it performs the following. All the untrue cases return an error message.
> > 
> > * checks if the number of arguments are correct (3)
> > * checks if the first argument contains either of the "load" or "save" keyword and if it does, creates a directory "/var/lib/systemd/backlight" if it doesn't already exist.
> > * checks the existence of ":" marker in the second argument and saves the name of the backlight device (which is given after ":" marker)
> > * checks if the seconds argument contains either of the "backlight" or "leds" keyword (type of the backlight device)
> > * checks if the device exists (by searching for it in "/sys/subsystem", "/sys/bus/", "/sys/class" directories)
> > * checks if the maximum brightness value of the backlight device is less than zero and if equal to zero. If neither of them are true, proceeds to the next step.
> > * If the first argument is "load", reads the brightness value of the device from "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file. If clamp option is true, it clamps (reduces) the read value. Sets the device's brightness attribute to the read value.
> > * If the first argument is "save", gets the brightness attribute of the device and saves the value to the "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file.
> > 
> > Thank you,
> 
> Thanks, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.

Thanks for the ack! After studying the tasks, I think I'll start with the task
"update outdated executables/daemons to new-style sd-bus macros".

**Comment by cerebro1** [comment]
Hi everyone,

Here is the screenshot of my first repeatable task :

![Screenshot from 2022-04-07 12-48-52](https://user-images.githubusercontent.com/10183232/162142833-2cfa89ed-5cef-43ec-818c-7ddfbde01ced.png)

**Comment by cerebro1** [comment]
Hi everyone,
Here is repeatable task 1 screenshot attached.
Repeatable Task 1 says to include log message so sharing screenshot with log message added for loginctl.
![Screenshot from 2022-04-07 13-01-25](https://user-images.githubusercontent.com/10183232/162145283-d3cead2a-5c58-4090-a7fc-1c00b497b2e3.png)

Since i have seen output of others already so that confused me whether output message was to be updated or log so to be on the safer side considering both have to be updated, here is the screenshot with output updated for tool loginctl.
![Screenshot from 2022-04-07 13-11-31](https://user-images.githubusercontent.com/10183232/162146602-ccc76425-3fc0-4228-a531-1fe1061ba1a1.png)

**Comment by cerebro1** [comment]
Hi everyone,

Here is the screenshot of Repeatable Task 1.
The task says to include log message so screenshot 1 shows syslog with log from loginctl
Screenshot 1
![Screenshot from 2022-04-07 13-56-06](https://user-images.githubusercontent.com/10183232/162155238-e0d3176b-c2a7-4a2b-a690-4dcdad868ec8.png)

Since i got confused that whether we need to update the STDOUT or syslog so here is screenshot 2 with STDOUT updated with expected message.
Screeshot 2
![Screenshot from 2022-04-07 14-00-00](https://user-images.githubusercontent.com/10183232/162156184-6e5b165b-fac1-4de0-a318-2cbd61634b53.png)

**Comment by bluca** [comment]
Thanks, looks great

**Comment by amarjargal** [comment]
> > > Hello everyone, For repeatable task 2, I chose the tool backlight. Running and inspecting the code, I tried to explain below what it does.
> > > When systemd-backlight tool is run, it first checks if the argument contains "help" keyword, it prints the manual for the tool. Then it performs the following. All the untrue cases return an error message.
> > > 
> > > * checks if the number of arguments are correct (3)
> > > * checks if the first argument contains either of the "load" or "save" keyword and if it does, creates a directory "/var/lib/systemd/backlight" if it doesn't already exist.
> > > * checks the existence of ":" marker in the second argument and saves the name of the backlight device (which is given after ":" marker)
> > > * checks if the seconds argument contains either of the "backlight" or "leds" keyword (type of the backlight device)
> > > * checks if the device exists (by searching for it in "/sys/subsystem", "/sys/bus/", "/sys/class" directories)
> > > * checks if the maximum brightness value of the backlight device is less than zero and if equal to zero. If neither of them are true, proceeds to the next step.
> > > * If the first argument is "load", reads the brightness value of the device from "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file. If clamp option is true, it clamps (reduces) the read value. Sets the device's brightness attribute to the read value.
> > > * If the first argument is "save", gets the brightness attribute of the device and saves the value to the "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file.
> > > 
> > > Thank you,
> > 
> > 
> > Thanks, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.
> 
> Thanks for the ack! After studying the tasks, I think I'll start with the task "update outdated executables/daemons to new-style sd-bus macros".

I updated machined service to use the new macros and opened a pull request. I followed the new contributors documentations, but since it's my first PR to systemd, I hope I didn't miss anything and everything is alright.

**Comment by bluca** [comment]
> > > > Hello everyone, For repeatable task 2, I chose the tool backlight. Running and inspecting the code, I tried to explain below what it does.
> > > > When systemd-backlight tool is run, it first checks if the argument contains "help" keyword, it prints the manual for the tool. Then it performs the following. All the untrue cases return an error message.
> > > > 
> > > > * checks if the number of arguments are correct (3)
> > > > * checks if the first argument contains either of the "load" or "save" keyword and if it does, creates a directory "/var/lib/systemd/backlight" if it doesn't already exist.
> > > > * checks the existence of ":" marker in the second argument and saves the name of the backlight device (which is given after ":" marker)
> > > > * checks if the seconds argument contains either of the "backlight" or "leds" keyword (type of the backlight device)
> > > > * checks if the device exists (by searching for it in "/sys/subsystem", "/sys/bus/", "/sys/class" directories)
> > > > * checks if the maximum brightness value of the backlight device is less than zero and if equal to zero. If neither of them are true, proceeds to the next step.
> > > > * If the first argument is "load", reads the brightness value of the device from "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file. If clamp option is true, it clamps (reduces) the read value. Sets the device's brightness attribute to the read value.
> > > > * If the first argument is "save", gets the brightness attribute of the device and saves the value to the "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file.
> > > > 
> > > > Thank you,
> > > 
> > > 
> > > Thanks, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.
> > 
> > 
> > Thanks for the ack! After studying the tasks, I think I'll start with the task "update outdated executables/daemons to new-style sd-bus macros".
> 
> I updated machined service to use the new macros and opened a pull request. I followed the new contributors documentations, but since it's my first PR to systemd, I hope I didn't miss anything and everything is alright.

It was great, thank you, here's the link for the record: https://github.com/systemd/systemd/pull/23012

**Comment by cerebro1** [comment]
Here is my repeatable task 2 : 
I have selected ac-power tool. Following is the explanation of code segement. Below that i have explained the complete executional flow of tool.

**Code Segment Explained :** 
- Include getopt.h from global and main-func.h and udev-util.h from current path
- Define static boolean variable arg_verbose and initialize to False. Flag to control if output as verbose.
- Define static function help. This function will print help message on STDOUT
- Define static parse_argv integer type function. This function will parse the input.
     - Create local enum with ARG_VERSION = 0x100
     - Create a static constant array options of option type and initialize it with help, version, verbose and one empty option.
     - Define an integer c.
     - Check argc is greater or equal to zero
     - Check argv is not NULL
     - Start a while loop with below condition:
           - Call getopt_long with argc, argv, options array.
           - Get return value of getopt_long in c
           - On each successive loop iteration, c is set to the next option character as defined in options array
           - If c is negative, exit the loop.
           - A Switch-case loop to process each option.
               - h : call help function print message
               - ARG_VERSION : call version function and STDOUT package version
               - v : set arg_verbose to true
               - ? : invalid option
               - In default case, function assert_not_reached is called to assure default case is never run.
           - If argument count passed is greater than expected for ac-power ( which is 0), return error that “systemd-ac-power takes no arguments”
     - parse_agrv  function parse the option and store the return value in r

**Code Executional Flow :**
- When the program runs, the macros inline and main-func.c executes. If ret > = 0, then the run function executes. It calls the parse_argv function to parse options and store return value in r. If r is negative, EXIT_FAILURE.
- Now check whether the external power source is plugged in via on_ac_power function. Return value stored in r. If r is negative, log error.
- If arg_verbose is true(when option passed is “-v”), call function yes_no and assign 1 as “no” for not plugged external power and 0 as “yes” for plugged.
- If this script is called with no option, it returns the calling method 0 if plugged external source and 1 if not plugged external source.

**Screeshot of execution :** 
![Screenshot from 2022-04-08 13-14-20](https://user-images.githubusercontent.com/10183232/162613948-ebf005f1-47dd-430a-a33e-0b7fd8c2347f.png)

**Comment by swarnimasaxena** [comment]
Hey everyone,
For repeatable task 3, I have chosen the manual page explaining file hierarchy system used by operating systems using systemd.
At the topmost level, there is the root directory denoted by '/'. It contains /boot/ directory which has the boot partition for bringing up the systems, the /etc/ directory which has system specific configuration, /home/ directory for normal user's home directories, /root/ directory which is the home directory of root user, /srv/ to store general server payload, and /tmp/ for storing small temporary files. It also contains directories storing runtime data like /run/ which is a "tmpfs" file system for system packages to place runtime data, socket files and similar, /run/log/ which stores runtime system logs and /run/user/ which contains per-user runtime directories.
Then we have the /usr/ directory which has the vendor-supplied operating system resources. It contains bin directory which stores the binaries and executables for user commands, include directory storing C and C++ API header files of system libraries, lib directory which includes internal executables or other binaries that are not regularly invoked from a shell, and share directory containing resources shared between multiple packages such as documentation, man pages etc. 
There is a /var/ directory storing persistent, variable system data such as system cache data, system data, logs, spool data, larger temporary files under different directories.
/dev/ is the root directory for device nodes, also containing a shm directory to store POSIX shared memory segments. Then there are /proc/ and /sys/ which are virtual kernel file systems exposing process list and discovered devices respectively. There are compatibility symlinks like /bin/, /sbin/, /lib/, /var/run/ etc which ensure that scripts and binaries or programs correctly find their binaries, resources or runtime data. 
There's also a structure to be followed by user applications to place files and directories in user's home directory. Developers of system packages should follow strict rules when placing their files in the file system. Programs running in user context should follow strict rules when placing their own files in the user's home directory.

**Comment by swarnimasaxena** [comment]
![Screenshot from 2022-04-09 04-01-31](https://user-images.githubusercontent.com/68150718/162635434-3638d9ee-6098-47d2-a071-f4af34848917.png)
This is the screenshot of repeatable task 1.

**Comment by bluca** [comment]
> Here is my repeatable task 2 : I have selected ac-power tool. Following is the explanation of code segement. Below that i have explained the complete executional flow of tool.
> 
> **Code Segment Explained :**
> 
> * Include getopt.h from global and main-func.h and udev-util.h from current path
> * Define static boolean variable arg_verbose and initialize to False. Flag to control if output as verbose.
> * Define static function help. This function will print help message on STDOUT
> * Define static parse_argv integer type function. This function will parse the input.
>   
>   * Create local enum with ARG_VERSION = 0x100
>   * Create a static constant array options of option type and initialize it with help, version, verbose and one empty option.
>   * Define an integer c.
>   * Check argc is greater or equal to zero
>   * Check argv is not NULL
>   * Start a while loop with below condition:
>     - Call getopt_long with argc, argv, options array.
>     - Get return value of getopt_long in c
>     - On each successive loop iteration, c is set to the next option character as defined in options array
>     - If c is negative, exit the loop.
>     - A Switch-case loop to process each option.
>     - h : call help function print message
>     - ARG_VERSION : call version function and STDOUT package version
>     - v : set arg_verbose to true
>     - ? : invalid option
>     - In default case, function assert_not_reached is called to assure default case is never run.
>     - If argument count passed is greater than expected for ac-power ( which is 0), return error that “systemd-ac-power takes no arguments”
>   * parse_agrv  function parse the option and store the return value in r
> 
> **Code Executional Flow :**
> 
> * When the program runs, the macros inline and fetches main-func.h. If ret > = 0, then the run function executes. It calls the parse_argv function to parse options and store return value in r. If r is negative, EXIT_FAILURE.
> * Now check whether the external power source is plugged in via on_ac_power function. Return value stored in r. If r is negative, log error.
> * If arg_verbose is true(when option passed is “-v”), call function yes_no and assign 1 as “no” for not plugged external power and 0 as “yes” for plugged.
> * If this script is called with no option, it returns the calling method 0 if plugged external source and 1 if not plugged external source.
> 
> **Screeshot of execution :** ![Screenshot from 2022-04-08 13-14-20](https://user-images.githubusercontent.com/10183232/162613948-ebf005f1-47dd-430a-a33e-0b7fd8c2347f.png)

Thanks @cerebro1, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.

**Comment by bluca** [comment]
> ![Screenshot from 2022-04-09 04-01-31](https://user-images.githubusercontent.com/68150718/162635434-3638d9ee-6098-47d2-a071-f4af34848917.png) This is the screenshot of repeatable task 1.

@swarnimasaxena for task 1, we are expecting you to update the source code of a tool and rebuild it, and run it, to print the message - but from the screenshot you are simply running the `logger` program. Could you please try again, modifying hostnamectl instead?

**Comment by swarnimasaxena** [comment]
> > ![Screenshot from 2022-04-09 04-01-31](https://user-images.githubusercontent.com/68150718/162635434-3638d9ee-6098-47d2-a071-f4af34848917.png) This is the screenshot of repeatable task 1.
> 
> @swarnimasaxena for task 1, we are expecting you to update the source code of a tool and rebuild it, and run it, to print the message - but from the screenshot you are simply running the `logger` program. Could you please try again, modifying hostnamectl instead?

![Screenshot from 2022-04-11 04-08-29](https://user-images.githubusercontent.com/68150718/162643052-ed7fd70b-7e88-4e19-ba87-17e5f6831bd5.png)

Is this okay?

**Comment by bluca** [comment]
> this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.

Thanks @swarnimasaxena - this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.

**Comment by swarnimasaxena** [comment]
> > this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.
> 
> Thanks @swarnimasaxena - this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.

Thank you, will get started right away.

**Comment by cerebro1** [comment]
> 

Thanks! 
I have gone through short task list and selected short task 4 : update outdated executables/daemons to bus_call_method and bus_call_property. I will raise PR in a while.

**Comment by cerebro1** [comment]
> > Here is my repeatable task 2 : I have selected ac-power tool. Following is the explanation of code segement. Below that i have explained the complete executional flow of tool.
> > **Code Segment Explained :**
> > 
> > * Include getopt.h from global and main-func.h and udev-util.h from current path
> > * Define static boolean variable arg_verbose and initialize to False. Flag to control if output as verbose.
> > * Define static function help. This function will print help message on STDOUT
> > * Define static parse_argv integer type function. This function will parse the input.
> >   
> >   * Create local enum with ARG_VERSION = 0x100
> >   * Create a static constant array options of option type and initialize it with help, version, verbose and one empty option.
> >   * Define an integer c.
> >   * Check argc is greater or equal to zero
> >   * Check argv is not NULL
> >   * Start a while loop with below condition:
> >     
> >     * Call getopt_long with argc, argv, options array.
> >     * Get return value of getopt_long in c
> >     * On each successive loop iteration, c is set to the next option character as defined in options array
> >     * If c is negative, exit the loop.
> >     * A Switch-case loop to process each option.
> >     * h : call help function print message
> >     * ARG_VERSION : call version function and STDOUT package version
> >     * v : set arg_verbose to true
> >     * ? : invalid option
> >     * In default case, function assert_not_reached is called to assure default case is never run.
> >     * If argument count passed is greater than expected for ac-power ( which is 0), return error that “systemd-ac-power takes no arguments”
> >   * parse_agrv  function parse the option and store the return value in r
> > 
> > **Code Executional Flow :**
> > 
> > * When the program runs, the macros inline and fetches main-func.h. If ret > = 0, then the run function executes. It calls the parse_argv function to parse options and store return value in r. If r is negative, EXIT_FAILURE.
> > * Now check whether the external power source is plugged in via on_ac_power function. Return value stored in r. If r is negative, log error.
> > * If arg_verbose is true(when option passed is “-v”), call function yes_no and assign 1 as “no” for not plugged external power and 0 as “yes” for plugged.
> > * If this script is called with no option, it returns the calling method 0 if plugged external source and 1 if not plugged external source.
> > 
> > **Screeshot of execution :** ![Screenshot from 2022-04-08 13-14-20](https://user-images.githubusercontent.com/10183232/162613948-ebf005f1-47dd-430a-a33e-0b7fd8c2347f.png)
> 
> Thanks @cerebro1, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.

@bluca I have raised PR for short task : update outdated executables/daemons to `bus_call_method` and `bus_call_property`

**Comment by amarjargal** [comment]
> > > > > Hello everyone, For repeatable task 2, I chose the tool backlight. Running and inspecting the code, I tried to explain below what it does.
> > > > > When systemd-backlight tool is run, it first checks if the argument contains "help" keyword, it prints the manual for the tool. Then it performs the following. All the untrue cases return an error message.
> > > > > 
> > > > > * checks if the number of arguments are correct (3)
> > > > > * checks if the first argument contains either of the "load" or "save" keyword and if it does, creates a directory "/var/lib/systemd/backlight" if it doesn't already exist.
> > > > > * checks the existence of ":" marker in the second argument and saves the name of the backlight device (which is given after ":" marker)
> > > > > * checks if the seconds argument contains either of the "backlight" or "leds" keyword (type of the backlight device)
> > > > > * checks if the device exists (by searching for it in "/sys/subsystem", "/sys/bus/", "/sys/class" directories)
> > > > > * checks if the maximum brightness value of the backlight device is less than zero and if equal to zero. If neither of them are true, proceeds to the next step.
> > > > > * If the first argument is "load", reads the brightness value of the device from "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file. If clamp option is true, it clamps (reduces) the read value. Sets the device's brightness attribute to the read value.
> > > > > * If the first argument is "save", gets the brightness attribute of the device and saves the value to the "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file.
> > > > > 
> > > > > Thank you,
> > > > 
> > > > 
> > > > Thanks, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.
> > > 
> > > 
> > > Thanks for the ack! After studying the tasks, I think I'll start with the task "update outdated executables/daemons to new-style sd-bus macros".
> > 
> > 
> > I updated machined service to use the new macros and opened a pull request. I followed the new contributors documentations, but since it's my first PR to systemd, I hope I didn't miss anything and everything is alright.
> 
> It was great, thank you, here's the link for the record: #23012

I created another PR for this task. If this PR is alright, then I'd like to move on to one of the unit test tasks.

**Comment by anitazha** [comment]
> > > Here is my repeatable task 2 : I have selected ac-power tool. Following is the explanation of code segement. Below that i have explained the complete executional flow of tool.
> > > **Code Segment Explained :**
> > > 
> > > * Include getopt.h from global and main-func.h and udev-util.h from current path
> > > * Define static boolean variable arg_verbose and initialize to False. Flag to control if output as verbose.
> > > * Define static function help. This function will print help message on STDOUT
> > > * Define static parse_argv integer type function. This function will parse the input.
> > >   
> > >   * Create local enum with ARG_VERSION = 0x100
> > >   * Create a static constant array options of option type and initialize it with help, version, verbose and one empty option.
> > >   * Define an integer c.
> > >   * Check argc is greater or equal to zero
> > >   * Check argv is not NULL
> > >   * Start a while loop with below condition:
> > >     
> > >     * Call getopt_long with argc, argv, options array.
> > >     * Get return value of getopt_long in c
> > >     * On each successive loop iteration, c is set to the next option character as defined in options array
> > >     * If c is negative, exit the loop.
> > >     * A Switch-case loop to process each option.
> > >     * h : call help function print message
> > >     * ARG_VERSION : call version function and STDOUT package version
> > >     * v : set arg_verbose to true
> > >     * ? : invalid option
> > >     * In default case, function assert_not_reached is called to assure default case is never run.
> > >     * If argument count passed is greater than expected for ac-power ( which is 0), return error that “systemd-ac-power takes no arguments”
> > >   * parse_agrv  function parse the option and store the return value in r
> > > 
> > > **Code Executional Flow :**
> > > 
> > > * When the program runs, the macros inline and fetches main-func.h. If ret > = 0, then the run function executes. It calls the parse_argv function to parse options and store return value in r. If r is negative, EXIT_FAILURE.
> > > * Now check whether the external power source is plugged in via on_ac_power function. Return value stored in r. If r is negative, log error.
> > > * If arg_verbose is true(when option passed is “-v”), call function yes_no and assign 1 as “no” for not plugged external power and 0 as “yes” for plugged.
> > > * If this script is called with no option, it returns the calling method 0 if plugged external source and 1 if not plugged external source.
> > > 
> > > **Screeshot of execution :** ![Screenshot from 2022-04-08 13-14-20](https://user-images.githubusercontent.com/10183232/162613948-ebf005f1-47dd-430a-a33e-0b7fd8c2347f.png)
> > 
> > 
> > Thanks @cerebro1, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.
> 
> @bluca I have raised PR for short task : update outdated executables/daemons to `bus_call_method` and `bus_call_property`

Thanks @cerebro1 for completion of your short task. Your PR https://github.com/systemd/systemd/pull/23047 has been merged.

**Comment by anitazha** [comment]
> > > > > > Hello everyone, For repeatable task 2, I chose the tool backlight. Running and inspecting the code, I tried to explain below what it does.
> > > > > > When systemd-backlight tool is run, it first checks if the argument contains "help" keyword, it prints the manual for the tool. Then it performs the following. All the untrue cases return an error message.
> > > > > > 
> > > > > > * checks if the number of arguments are correct (3)
> > > > > > * checks if the first argument contains either of the "load" or "save" keyword and if it does, creates a directory "/var/lib/systemd/backlight" if it doesn't already exist.
> > > > > > * checks the existence of ":" marker in the second argument and saves the name of the backlight device (which is given after ":" marker)
> > > > > > * checks if the seconds argument contains either of the "backlight" or "leds" keyword (type of the backlight device)
> > > > > > * checks if the device exists (by searching for it in "/sys/subsystem", "/sys/bus/", "/sys/class" directories)
> > > > > > * checks if the maximum brightness value of the backlight device is less than zero and if equal to zero. If neither of them are true, proceeds to the next step.
> > > > > > * If the first argument is "load", reads the brightness value of the device from "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file. If clamp option is true, it clamps (reduces) the read value. Sets the device's brightness attribute to the read value.
> > > > > > * If the first argument is "save", gets the brightness attribute of the device and saves the value to the "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file.
> > > > > > 
> > > > > > Thank you,
> > > > > 
> > > > > 
> > > > > Thanks, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.
> > > > 
> > > > 
> > > > Thanks for the ack! After studying the tasks, I think I'll start with the task "update outdated executables/daemons to new-style sd-bus macros".
> > > 
> > > 
> > > I updated machined service to use the new macros and opened a pull request. I followed the new contributors documentations, but since it's my first PR to systemd, I hope I didn't miss anything and everything is alright.
> > 
> > 
> > It was great, thank you, here's the link for the record: #23012
> 
> I created another PR for this task. If this PR is alright, then I'd like to move on to one of the unit test tasks.

Thanks @amarjargal for completing your short task and getting #23012 merged. You're free to work on a unit test if you'd like.

**Comment by amarjargal** [comment]
> > > > > > > Hello everyone, For repeatable task 2, I chose the tool backlight. Running and inspecting the code, I tried to explain below what it does.
> > > > > > > When systemd-backlight tool is run, it first checks if the argument contains "help" keyword, it prints the manual for the tool. Then it performs the following. All the untrue cases return an error message.
> > > > > > > 
> > > > > > > * checks if the number of arguments are correct (3)
> > > > > > > * checks if the first argument contains either of the "load" or "save" keyword and if it does, creates a directory "/var/lib/systemd/backlight" if it doesn't already exist.
> > > > > > > * checks the existence of ":" marker in the second argument and saves the name of the backlight device (which is given after ":" marker)
> > > > > > > * checks if the seconds argument contains either of the "backlight" or "leds" keyword (type of the backlight device)
> > > > > > > * checks if the device exists (by searching for it in "/sys/subsystem", "/sys/bus/", "/sys/class" directories)
> > > > > > > * checks if the maximum brightness value of the backlight device is less than zero and if equal to zero. If neither of them are true, proceeds to the next step.
> > > > > > > * If the first argument is "load", reads the brightness value of the device from "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file. If clamp option is true, it clamps (reduces) the read value. Sets the device's brightness attribute to the read value.
> > > > > > > * If the first argument is "save", gets the brightness attribute of the device and saves the value to the "/var/lib/systemd/backlight/TYPE_OF_DEVICE:NAME_OF_DEVICE" file.
> > > > > > > 
> > > > > > > Thank you,
> > > > > > 
> > > > > > 
> > > > > > Thanks, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.
> > > > > 
> > > > > 
> > > > > Thanks for the ack! After studying the tasks, I think I'll start with the task "update outdated executables/daemons to new-style sd-bus macros".
> > > > 
> > > > 
> > > > I updated machined service to use the new macros and opened a pull request. I followed the new contributors documentations, but since it's my first PR to systemd, I hope I didn't miss anything and everything is alright.
> > > 
> > > 
> > > It was great, thank you, here's the link for the record: #23012
> > 
> > 
> > I created another PR for this task. If this PR is alright, then I'd like to move on to one of the unit test tasks.
> 
> Thanks @amarjargal for completing your short task and getting #23012 merged. You're free to work on a unit test if you'd like.

Thank you! I also have #23049 merged. Now studying the other tasks.

**Comment by cerebro1** [comment]
> > > > Here is my repeatable task 2 : I have selected ac-power tool. Following is the explanation of code segement. Below that i have explained the complete executional flow of tool.
> > > > **Code Segment Explained :**
> > > > 
> > > > * Include getopt.h from global and main-func.h and udev-util.h from current path
> > > > * Define static boolean variable arg_verbose and initialize to False. Flag to control if output as verbose.
> > > > * Define static function help. This function will print help message on STDOUT
> > > > * Define static parse_argv integer type function. This function will parse the input.
> > > >   
> > > >   * Create local enum with ARG_VERSION = 0x100
> > > >   * Create a static constant array options of option type and initialize it with help, version, verbose and one empty option.
> > > >   * Define an integer c.
> > > >   * Check argc is greater or equal to zero
> > > >   * Check argv is not NULL
> > > >   * Start a while loop with below condition:
> > > >     
> > > >     * Call getopt_long with argc, argv, options array.
> > > >     * Get return value of getopt_long in c
> > > >     * On each successive loop iteration, c is set to the next option character as defined in options array
> > > >     * If c is negative, exit the loop.
> > > >     * A Switch-case loop to process each option.
> > > >     * h : call help function print message
> > > >     * ARG_VERSION : call version function and STDOUT package version
> > > >     * v : set arg_verbose to true
> > > >     * ? : invalid option
> > > >     * In default case, function assert_not_reached is called to assure default case is never run.
> > > >     * If argument count passed is greater than expected for ac-power ( which is 0), return error that “systemd-ac-power takes no arguments”
> > > >   * parse_agrv  function parse the option and store the return value in r
> > > > 
> > > > **Code Executional Flow :**
> > > > 
> > > > * When the program runs, the macros inline and fetches main-func.h. If ret > = 0, then the run function executes. It calls the parse_argv function to parse options and store return value in r. If r is negative, EXIT_FAILURE.
> > > > * Now check whether the external power source is plugged in via on_ac_power function. Return value stored in r. If r is negative, log error.
> > > > * If arg_verbose is true(when option passed is “-v”), call function yes_no and assign 1 as “no” for not plugged external power and 0 as “yes” for plugged.
> > > > * If this script is called with no option, it returns the calling method 0 if plugged external source and 1 if not plugged external source.
> > > > 
> > > > **Screeshot of execution :** ![Screenshot from 2022-04-08 13-14-20](https://user-images.githubusercontent.com/10183232/162613948-ebf005f1-47dd-430a-a33e-0b7fd8c2347f.png)
> > > 
> > > 
> > > Thanks @cerebro1, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.
> > 
> > 
> > @bluca I have raised PR for short task : update outdated executables/daemons to `bus_call_method` and `bus_call_property`
> 
> Thanks @cerebro1 for completion of your short task. Since your PR #23047 has been merged you're free to work on a long task if you'd like.

Thanks. I have gone through the list and selected an issue labelled as RFE for option in systemctl to create new unit via prepopulated template file.

**Comment by anitazha** [comment]
> > > > > Here is my repeatable task 2 : I have selected ac-power tool. Following is the explanation of code segement. Below that i have explained the complete executional flow of tool.
> > > > > **Code Segment Explained :**
> > > > > 
> > > > > * Include getopt.h from global and main-func.h and udev-util.h from current path
> > > > > * Define static boolean variable arg_verbose and initialize to False. Flag to control if output as verbose.
> > > > > * Define static function help. This function will print help message on STDOUT
> > > > > * Define static parse_argv integer type function. This function will parse the input.
> > > > >   
> > > > >   * Create local enum with ARG_VERSION = 0x100
> > > > >   * Create a static constant array options of option type and initialize it with help, version, verbose and one empty option.
> > > > >   * Define an integer c.
> > > > >   * Check argc is greater or equal to zero
> > > > >   * Check argv is not NULL
> > > > >   * Start a while loop with below condition:
> > > > >     
> > > > >     * Call getopt_long with argc, argv, options array.
> > > > >     * Get return value of getopt_long in c
> > > > >     * On each successive loop iteration, c is set to the next option character as defined in options array
> > > > >     * If c is negative, exit the loop.
> > > > >     * A Switch-case loop to process each option.
> > > > >     * h : call help function print message
> > > > >     * ARG_VERSION : call version function and STDOUT package version
> > > > >     * v : set arg_verbose to true
> > > > >     * ? : invalid option
> > > > >     * In default case, function assert_not_reached is called to assure default case is never run.
> > > > >     * If argument count passed is greater than expected for ac-power ( which is 0), return error that “systemd-ac-power takes no arguments”
> > > > >   * parse_agrv  function parse the option and store the return value in r
> > > > > 
> > > > > **Code Executional Flow :**
> > > > > 
> > > > > * When the program runs, the macros inline and fetches main-func.h. If ret > = 0, then the run function executes. It calls the parse_argv function to parse options and store return value in r. If r is negative, EXIT_FAILURE.
> > > > > * Now check whether the external power source is plugged in via on_ac_power function. Return value stored in r. If r is negative, log error.
> > > > > * If arg_verbose is true(when option passed is “-v”), call function yes_no and assign 1 as “no” for not plugged external power and 0 as “yes” for plugged.
> > > > > * If this script is called with no option, it returns the calling method 0 if plugged external source and 1 if not plugged external source.
> > > > > 
> > > > > **Screeshot of execution :** ![Screenshot from 2022-04-08 13-14-20](https://user-images.githubusercontent.com/10183232/162613948-ebf005f1-47dd-430a-a33e-0b7fd8c2347f.png)
> > > > 
> > > > 
> > > > Thanks @cerebro1, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.
> > > 
> > > 
> > > @bluca I have raised PR for short task : update outdated executables/daemons to `bus_call_method` and `bus_call_property`
> > 
> > 
> > Thanks @cerebro1 for completion of your short task. Since your PR #23047 has been merged you're free to work on a long task if you'd like.
> 
> Thanks. I have gone through the list and selected an issue labelled as RFE for option in systemctl to create new unit via prepopulated template file.

Please link the issue number :) This will help us better advise

**Comment by cerebro1** [comment]
> > > > > > Here is my repeatable task 2 : I have selected ac-power tool. Following is the explanation of code segement. Below that i have explained the complete executional flow of tool.
> > > > > > **Code Segment Explained :**
> > > > > > 
> > > > > > * Include getopt.h from global and main-func.h and udev-util.h from current path
> > > > > > * Define static boolean variable arg_verbose and initialize to False. Flag to control if output as verbose.
> > > > > > * Define static function help. This function will print help message on STDOUT
> > > > > > * Define static parse_argv integer type function. This function will parse the input.
> > > > > >   
> > > > > >   * Create local enum with ARG_VERSION = 0x100
> > > > > >   * Create a static constant array options of option type and initialize it with help, version, verbose and one empty option.
> > > > > >   * Define an integer c.
> > > > > >   * Check argc is greater or equal to zero
> > > > > >   * Check argv is not NULL
> > > > > >   * Start a while loop with below condition:
> > > > > >     
> > > > > >     * Call getopt_long with argc, argv, options array.
> > > > > >     * Get return value of getopt_long in c
> > > > > >     * On each successive loop iteration, c is set to the next option character as defined in options array
> > > > > >     * If c is negative, exit the loop.
> > > > > >     * A Switch-case loop to process each option.
> > > > > >     * h : call help function print message
> > > > > >     * ARG_VERSION : call version function and STDOUT package version
> > > > > >     * v : set arg_verbose to true
> > > > > >     * ? : invalid option
> > > > > >     * In default case, function assert_not_reached is called to assure default case is never run.
> > > > > >     * If argument count passed is greater than expected for ac-power ( which is 0), return error that “systemd-ac-power takes no arguments”
> > > > > >   * parse_agrv  function parse the option and store the return value in r
> > > > > > 
> > > > > > **Code Executional Flow :**
> > > > > > 
> > > > > > * When the program runs, the macros inline and fetches main-func.h. If ret > = 0, then the run function executes. It calls the parse_argv function to parse options and store return value in r. If r is negative, EXIT_FAILURE.
> > > > > > * Now check whether the external power source is plugged in via on_ac_power function. Return value stored in r. If r is negative, log error.
> > > > > > * If arg_verbose is true(when option passed is “-v”), call function yes_no and assign 1 as “no” for not plugged external power and 0 as “yes” for plugged.
> > > > > > * If this script is called with no option, it returns the calling method 0 if plugged external source and 1 if not plugged external source.
> > > > > > 
> > > > > > **Screeshot of execution :** ![Screenshot from 2022-04-08 13-14-20](https://user-images.githubusercontent.com/10183232/162613948-ebf005f1-47dd-430a-a33e-0b7fd8c2347f.png)
> > > > > 
> > > > > 
> > > > > Thanks @cerebro1, this is the second initial task, so you can now move on to the short/long tasks. Let us know if you need any clarification.
> > > > 
> > > > 
> > > > @bluca I have raised PR for short task : update outdated executables/daemons to `bus_call_method` and `bus_call_property`
> > > 
> > > 
> > > Thanks @cerebro1 for completion of your short task. Since your PR #23047 has been merged you're free to work on a long task if you'd like.
> > 
> > 
> > Thanks. I have gone through the list and selected an issue labelled as RFE for option in systemctl to create new unit via prepopulated template file.
> 
> Please link the issue number :) This will help us better advise

Hi @anitazha ,
I have been going through this issue. https://github.com/systemd/systemd/issues/22826

**Comment by amarjargal** [comment]
I'm working on issue #19256. I think I can manage this by studying the start, stop and is-active commands and doing something similar to them.

**Comment by anitazha** [comment]
> 

I would probably start on the C parts and see how far you get. The base part is relatively simple: new `systemctl` argument that copies a file to a new directory under a different name. Then if you have time you can add additional features and hook ups to create the templates directory.

So for example, the base code would do the following:
- add a new `systemctl --new` argument that requires the user to pass in the name of a unit (e.g. "test.service")
- if `/usr/share/systemd/unit/template` exists, open/copy the template file corresponding to the unit name  into `/etc/systemd/system` with the supplied unit name (e.g. "service" template since "test.service" is the example name).
- if the template doesn't exist or the name conflicts with an existing unit, you can return an error code.

Then to test it, you can manually create `/usr/share/systemd/unit/template` in your test environment and populate it with a test service unit template.

If you have time you can start creating templates and adding the meson parts.

**Comment by anitazha** [comment]
@cerebro1 sorry forgot to tag. see suggestion on https://github.com/systemd/systemd/issues/22192#issuecomment-1098495357 possibly putting up a commit with a subset of the features and see how far you get

**Comment by bluca** [comment]
Reminder for all applicants from Outreachy:

> The deadline for final applications is April 22, 2022 at 4pm UTC. Applicants will not be able to create a final application after the deadline. Applicants can edit their final application until the deadline.

**Comment by bluca** [comment]
Another reminder: final applications must be submitted by Friday (4pm UTC). It seems the project description might not be public yet for some reason (or at least I can't access this page without logging in:
https://www.outreachy.org/outreachy-may-2022-internship-round/communities/systemd/
)

The detailed tasks, which the final application form asks about, are listed under this RFE, which is what will be
used for the project:

https://github.com/systemd/systemd/issues/21165

**Comment by cerebro1** [comment]
@bluca @anitazha I was going through TODO and found this https://github.com/systemd/systemd/blob/main/TODO#L140
Do i need to open an issue for this or directly raise PR once i have made relevant changes?

**Comment by bluca** [comment]
Just create a PR

**Comment by amarjargal** [comment]
> Another reminder: final applications must be submitted by Friday (4pm UTC). It seems the project description might not be public yet for some reason (or at least I can't access this page without logging in: https://www.outreachy.org/outreachy-may-2022-internship-round/communities/systemd/ )
> 
> The detailed tasks, which the final application form asks about, are listed under this RFE, which is what will be used for the project:
> 
> #21165

I'm working on my final application. There is a field about community specific question and they advise to double check with our mentors about it, if there is any additional information that might be provided.

**Comment by bluca** [comment]
> > Another reminder: final applications must be submitted by Friday (4pm UTC). It seems the project description might not be public yet for some reason (or at least I can't access this page without logging in: https://www.outreachy.org/outreachy-may-2022-internship-round/communities/systemd/ )
> > The detailed tasks, which the final application form asks about, are listed under this RFE, which is what will be used for the project:
> > #21165
> 
> I'm working on my final application. There is a field about community specific question and they advise to double check with our mentors about it, if there is any additional information that might be provided.

Nothing specific comes to mind, that is not already mentioned in the project page or here

**Comment by bluca** [comment]
The application period is over, and the Outreachy org will reach out to applicants on the 20th of May. Thank you everyone for your participation and contributions!
